src/dnaio/_core.pyx

```diff
@@ -258,122 +266,129 @@ def fastq_iter(file, sequence_class, Py_ssize_t buffer_size):
                 bufstart += 1
                 bufend += 1
                 extra_newline = True
-            else:
-                break
+            elif last_read_position > record_start:  # Incomplete FASTQ records are present.
+                if extra_newline:
+                    # Do not report the linefeed that was added by dnaio but
+                    # was not present in the original input.
+                    last_read_position -= 1
+                lines = buf[record_start:last_read_position].count(b'\n')
+                raise FastqFormatError(
+                    'Premature end of file encountered. The incomplete final record was: '
+                    '{!r}'.format(
+                        shorten(buf[record_start:last_read_position].decode('latin-1'),
+                                500)),
+                    line=n_records * 4 + lines)
+            else:  # EOF Reached. Stop iterating.
+                return
 
         # Parse all complete FASTQ records in this chunk
-        pos = 0
         record_start = 0
-        while True:
-            # Parse the name (line 0)
-            if c_buf[pos] != b'@':
-                raise FastqFormatError("Line expected to "
-                    "start with '@', but found {!r}".format(chr(c_buf[pos])),
-                    line=n_records * 4)
-            pos += 1
-            while pos < bufend and c_buf[pos] != b'\n':
-                pos += 1
-            if pos == bufend:
+        while record_start < bufend:
+            ### Check for a complete record (i.e 4 newlines are present)
+            # Use libc memchr, this optimizes looking for characters by
+            # using 64-bit integers. See:
+            # https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=string/memchr.c;hb=HEAD
+            # void *memchr(const void *str, int c, size_t n)
+            name_end_ptr = <char *>memchr(c_buf + record_start, 10, <size_t>(bufend - record_start))
+            if name_end_ptr == NULL or name_end_ptr == bufend_ptr:
```

Comment:
Can `name_end_ptr == bufend_ptr` ever be True? `bufend_ptr` points to the byte just after the buffer and I wouldn’t expect `memchr` to ever look at this.

Reply:
Indeed. It points at the null byte so it is never true. Well-spotted! I can remove those checks and everything seems fine.

The code will never crash because. 
- If there is no linefeed it returns NULL
- If there is a linefeed at the last character, `name_end` will be at `bufend -1`. It can never be at `bufend` because that is a \00 byte. In that case, if we increase with +1  the next point will be at bufend. Bufend-bufend == 0. When memchr gets a size 0, it will return NULL. So it is safe to do a pointer increase with +1 inbetween memchr calls. But not more than one.

I added these snippets because of `if pos == bufend`, and the code worked fine with no crashes. But thinking about it, they are not necessary.

Reply:
I adopted the code. It also means that testing next_record_start at `sequence_end + 1`. `bufend` is the maximum value and in that case memchr will return NULL.